developer/frameworks/articles/290-configuration/09-fully-custom-system-settings-ui.markdown

```diff
@@ -0,0 +1,292 @@
+to to convert a map of the configuration's properties to a typed class. The
+configuration is stored in a volatile field. Don't forget to make it volatile
+to prevent thread safety problems. See the article on
+[reading configuration values from a component class](/docs/7-2/frameworks/-/knowledge_base/f/reading-configuration-values-from-a-component)
+for more information.
+
+```java
+
+@Activate
+@Modified
+public void activate(Map<String, Object> properties) {
+    _siteNavigationLanguageWebTemplateConfiguration =
+        ConfigurableUtil.createConfigurable(
+            SiteNavigationLanguageWebTemplateConfiguration.class,
+            properties);
+}
+
+private volatile SiteNavigationLanguageWebTemplateConfiguration
+    _siteNavigationLanguageWebTemplateConfiguration;
+
+```
+
+Next override the `getPid` and `getRequestParameters` methods:
+
+```java
+
+@Override
+public String getPid() {
+    return "com.liferay.site.navigation.language.web.configuration." +
+        "SiteNavigationLanguageWebTemplateConfiguration";
+}
+
+```
+
+Return the full configuration ID, as specified in the `*Configuration` class's
+`@Meta.OCD` annotation.
+
+```java
+
+@Override
+public Map<String, Object> getRequestParameters(
+    HttpServletRequest request) {
+
+    Map<String, Object> params = new HashMap<>();
+
+    String ddmTemplateKey = ParamUtil.getString(request, "ddmTemplateKey");
+
+    params.put("ddmTemplateKey", ddmTemplateKey);
+
+    return params;
+}
+
+```
+
+In the `getRequestParameters` method, map the parameters sent by the custom form
+(obtained from the request) to the keys of the fields in the Configuration
+interface.
+
+Provide the render logic via the overridden `render` method. The rendering
+approach demonstrated here uses a JSP. Recall that it's backed by a
+`DisplayContext` class set into the request object. The values set from this
+`render` method are available in the JSP via the `DisplayContext` object's
+getters. 
+
+Loop through the DDM Template Keys for the given `groupId` and set them into
+the display context with the `addTemplateKey` method. Then set the other
+necessary values that the JSP needs. In this case, set the title, the field
+label, and the redirect URL. Finally, call `renderJSP` and pass in the
+`servletContext`, request, response, and the path to the JSP: 
+
+```java
+
+@Override
+public void render(HttpServletRequest request, HttpServletResponse response)
+    throws IOException {
+
+    Locale locale = LocaleThreadLocal.getThemeDisplayLocale();
+
+    LanguageTemplateConfigurationDisplayContext
+        languageTemplateConfigurationDisplayContext =
+            new LanguageTemplateConfigurationDisplayContext();
+
+    languageTemplateConfigurationDisplayContext.setCurrentTemplateName(
+        _siteNavigationLanguageWebTemplateConfiguration.ddmTemplateKey());
+
+    long groupId = 0;
+
+    Group group = _groupLocalService.fetchCompanyGroup(
+        CompanyThreadLocal.getCompanyId());
```

Comment:
@drewbrokke 

I'm confused by your offer to help, because it doesn't look hard.
I replaced the lines 183 and 184 with the lines you gave me. Should I do anything else?

Reply:
No worries, that should do it.  👍